public/app/plugins/datasource/loki/datasource.ts

```diff
@@ -161,62 +161,82 @@ export class LokiDatasource
     return [SupplementaryQueryType.LogsVolume, SupplementaryQueryType.LogsSample];
   }
 
-  getLogsVolumeDataProvider(request: DataQueryRequest<LokiQuery>): Observable<DataQueryResponse> | undefined {
-    const isQuerySuitable = (query: LokiQuery) => {
-      const normalized = getNormalizedLokiQuery(query);
-      const { expr } = normalized;
-      // it has to be a logs-producing range-query
-      return expr && isLogsQuery(expr) && normalized.queryType === LokiQueryType.Range;
-    };
-
-    const isLogsVolumeAvailable = request.targets.some(isQuerySuitable);
-
-    if (!isLogsVolumeAvailable) {
+  getSupplementaryQuery(type: SupplementaryQueryType, query: LokiQuery): LokiQuery | undefined {
+    if (!this.getSupportedSupplementaryQueryTypes().includes(type)) {
       return undefined;
     }
 
-    const logsVolumeRequest = cloneDeep(request);
-    logsVolumeRequest.targets = logsVolumeRequest.targets.filter(isQuerySuitable).map((target) => {
-      const query = removeCommentsFromQuery(target.expr);
-      return {
-        ...target,
-        refId: `${REF_ID_STARTER_LOG_VOLUME}${target.refId}`,
-        instant: false,
-        volumeQuery: true,
-        expr: `sum by (level) (count_over_time(${query}[$__interval]))`,
-      };
-    });
+    const normalizedQuery = getNormalizedLokiQuery(query);
+    const expr = removeCommentsFromQuery(normalizedQuery.expr);
+    let isQuerySuitable = false;
+    let supplementaryQuery = undefined;
 
-    return queryLogsVolume(this, logsVolumeRequest, {
-      extractLevel,
-      range: request.range,
-      targets: request.targets,
-    });
-  }
+    switch (type) {
+      case SupplementaryQueryType.LogsVolume:
+        // it has to be a logs-producing range-query
+        isQuerySuitable = !!(query.expr && isLogsQuery(query.expr) && query.queryType === LokiQueryType.Range);
+        if (isQuerySuitable) {
+          supplementaryQuery = {
+            ...normalizedQuery,
+            refId: `${REF_ID_STARTER_LOG_VOLUME}${normalizedQuery.refId}`,
+            instant: false,
+            volumeQuery: true,
+            expr: `sum by (level) (count_over_time(${expr}[$__interval]))`,
+          };
+        }
+        break;
 
-  getLogsSampleDataProvider(request: DataQueryRequest<LokiQuery>): Observable<DataQueryResponse> | undefined {
-    const isQuerySuitable = (query: LokiQuery) => {
-      return query.expr && !isLogsQuery(query.expr);
-    };
+      case SupplementaryQueryType.LogsSample:
+        // it has to be a metric query
+        isQuerySuitable = !!(query.expr && !isLogsQuery(query.expr));
+        if (isQuerySuitable) {
+          supplementaryQuery = {
+            ...normalizedQuery,
+            refId: `${REF_ID_STARTER_LOG_SAMPLE}${normalizedQuery.refId}`,
+            expr: getLogQueryFromMetricsQuery(expr),
+            maxLines: 100,
+          };
+        }
+        break;
+
+      default:
+        supplementaryQuery = undefined;
```

Comment:
And.
```suggestion
```

Reply:
Fixed and refactored in https://github.com/grafana/grafana/pull/62097/commits/6bc2b4c9925c56d2260168076329133753fd2981?diff=split&w=1